src/vmm/src/builder.rs

```diff
@@ -763,39 +767,677 @@ pub mod tests {
     use std::fs::File;
 
     use super::*;
+    use arch::DeviceType;
+    use builder::StartMicrovmError::*;
+    use devices::virtio::TYPE_BLOCK;
+    use kernel::cmdline::Cmdline;
     use polly::event_manager::EventManager;
+    use seccomp::SECCOMP_LEVEL_ADVANCED;
+    use std::path::PathBuf;
+    use vmm_config::boot_source::DEFAULT_KERNEL_CMDLINE;
+    use vmm_config::drive::BlockDeviceConfig;
+    #[cfg(target_arch = "x86_64")]
+    use vmm_config::machine_config::CpuFeaturesTemplate;
+    use vmm_config::net::NetworkInterfaceConfig;
 
-    #[test]
-    fn test_stdin_wrapper() {
-        let wrapper = SerialStdin(io::stdin());
-        assert_eq!(wrapper.as_raw_fd(), io::stdin().as_raw_fd())
+    struct SerialInput(File);
+    impl io::Read for SerialInput {
+        fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+            self.0.read(buf)
+        }
     }
+    impl AsRawFd for SerialInput {
+        fn as_raw_fd(&self) -> RawFd {
+            self.0.as_raw_fd()
+        }
+    }
+    impl devices::legacy::ReadableFd for SerialInput {}
 
-    #[test]
-    fn test_setup_serial_device() {
-        // Wrapper over TempFile that implements `ReadableFd`.
-        struct SerialInput(File);
-        impl io::Read for SerialInput {
-            fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
-                self.0.read(buf)
-            }
+    fn default_kernel_file() -> File {
+        let path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+        let parent = path.parent().unwrap();
+
+        #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+        {
+            File::open(parent.join("kernel/src/loader/test_elf.bin")).unwrap()
         }
-        impl AsRawFd for SerialInput {
-            fn as_raw_fd(&self) -> RawFd {
-                self.0.as_raw_fd()
-            }
+
+        #[cfg(target_arch = "aarch64")]
+        {
+            File::open(parent.join("kernel/src/loader/test_pe.bin")).unwrap()
         }
-        impl devices::legacy::ReadableFd for SerialInput {};
+    }
+
+    fn default_mmio_device_manager() -> MMIODeviceManager {
+        MMIODeviceManager::new(
+            &mut (arch::MMIO_MEM_START as u64),
+            (arch::IRQ_BASE, arch::IRQ_MAX),
+        )
+    }
 
+    #[cfg(target_arch = "x86_64")]
+    fn default_portio_device_manager() -> PortIODeviceManager {
         let read_tempfile = TempFile::new("/tmp/serial_read_").unwrap();
         let read_file = File::open(read_tempfile.as_path()).unwrap();
         let read_handle = SerialInput(read_file);
         let mut event_manager = EventManager::new().expect("Unable to create EventManager");
-        setup_serial_device(
+
+        let serial = setup_serial_device(
```

Comment:
It is a bit weird for the Event Manager to be created here and then not be accessible any more. It should be created outside of this function and passed as a parameter.

Reply:
I agree. I've changed that. Take a look please.